src/graph.rs

```diff
@@ -5870,6 +5863,62 @@ mod tests {
     );
   }
 
+  #[tokio::test]
+  async fn dependency_ts_type_import_with_deno_types() {
+    struct TestLoader;
+    impl Loader for TestLoader {
+      fn load(
+        &self,
+        specifier: &ModuleSpecifier,
+        options: LoadOptions,
+      ) -> LoadFuture {
+        let specifier = specifier.clone();
+        match specifier.as_str() {
+          "file:///foo.ts" => Box::pin(async move {
+            Ok(Some(LoadResponse::Module {
+              specifier: specifier.clone(),
+              maybe_headers: None,
+              content: b"
+                // @deno-types='file:///bar.d.ts'
+                import type { Bar as _ } from 'bar';
+              "
+              .to_vec()
+              .into(),
+            }))
+          }),
+          "file:///bar.d.ts" => {
+            assert!(!options.is_dynamic);
+            Box::pin(async move {
+              Ok(Some(LoadResponse::Module {
+                specifier: specifier.clone(),
+                maybe_headers: None,
+                content: b"export type Bar = null;\n".to_vec().into(),
+              }))
+            })
+          }
+          _ => unreachable!(),
+        }
+      }
+    }
+    let mut graph = ModuleGraph::new(GraphKind::TypesOnly);
+    graph
+      .build(
+        vec![Url::parse("file:///foo.ts").unwrap()],
+        &TestLoader,
+        Default::default(),
+      )
+      .await;
+    graph.valid().unwrap();
+    let module = graph.get(&Url::parse("file:///foo.ts").unwrap()).unwrap();
+    let module = module.js().unwrap();
+    let dependency = module.dependencies.get("bar").unwrap();
+    assert_eq!(dependency.maybe_code, Resolution::None);
+    assert_eq!(
+      dependency.maybe_type.maybe_specifier().unwrap().as_str(),
+      "file:///bar.d.ts",
+    );
+  }
```

Comment:
Note for the future: it's more maintainable to add a test like this in tests/specs/graph probably.